lib/jxl/encode.cc

```diff
@@ -29,43 +33,364 @@
+      size_t size = requested_size + additional_size;
+      uint8_t* user_buffer =
+          static_cast<uint8_t*>(external_output_processor_->get_buffer(
+              external_output_processor_->opaque, &size));
+      if (size == 0 || user_buffer == nullptr) {
+        stop_requested_ = true;
+        return jxl::StatusCode::kNotEnoughBytes;
+      }
+      if (size < min_size + additional_size) {
+        external_output_processor_->release_buffer(
+            external_output_processor_->opaque, 0);
+      } else {
+        internal_buffers_.emplace(position_, InternalBuffer());
+        this->has_buffer_ = true;
+        return JxlOutputProcessorBuffer(user_buffer + additional_size,
+                                        size - additional_size, 0, this);
+      }
+    }
+  } else {
+    if (min_size + additional_size < *avail_out_) {
+      internal_buffers_.emplace(position_, InternalBuffer());
+      this->has_buffer_ = true;
+      return JxlOutputProcessorBuffer(*next_out_ + additional_size,
+                                      *avail_out_ - additional_size, 0, this);
+    }
+  }
+
+  // Otherwise, we need to allocate our own buffer.
+  auto it = internal_buffers_.emplace(position_, InternalBuffer()).first;
+  InternalBuffer& buffer = it->second;
+  size_t alloc_size = requested_size;
+  it++;
+  if (it != internal_buffers_.end()) {
+    alloc_size = std::min(alloc_size, it->first - position_);
+    JXL_ASSERT(alloc_size >= min_size);
+  }
+  buffer.owned_data.resize(alloc_size);
+  this->has_buffer_ = true;
+  return JxlOutputProcessorBuffer(buffer.owned_data.data(), alloc_size, 0,
+                                  this);
+}
+
+void JxlEncoderOutputProcessorWrapper::Seek(size_t pos) {
+  JXL_ASSERT(!has_buffer_);
+  if (external_output_processor_ && external_output_processor_->seek) {
+    external_output_processor_->seek(external_output_processor_->opaque, pos);
+    output_position_ = pos;
+  }
+  JXL_ASSERT(pos >= finalized_position_);
+  position_ = pos;
+}
+
+void JxlEncoderOutputProcessorWrapper::SetWatermark(size_t pos) {
+  JXL_ASSERT(!has_buffer_);
+  if (external_output_processor_ && external_output_processor_->seek) {
+    external_output_processor_->set_finalized_position(
+        external_output_processor_->opaque, pos);
+  }
+  finalized_position_ = pos;
+  FlushOutput();
+}
+
+bool JxlEncoderOutputProcessorWrapper::SetAvailOut(uint8_t** next_out,
+                                                   size_t* avail_out) {
+  if (external_output_processor_) return false;
+  avail_out_ = avail_out;
+  next_out_ = next_out;
+  FlushOutput();
+  return true;
+}
+
+void JxlEncoderOutputProcessorWrapper::ReleaseBuffer(size_t bytes_used) {
+  JXL_ASSERT(has_buffer_);
+  has_buffer_ = false;
+  auto it = internal_buffers_.find(position_);
+  if (it->second.owned_data.empty() && external_output_processor_) {
+    external_output_processor_->release_buffer(
+        external_output_processor_->opaque, bytes_used);
```

Comment:
shouldn't output_position_ be incremented by bytes_used here? otherwise it will not hold that output_position_ is the position of the external_output_processor_

Reply:
thanks makes sense. Changed it.

Reply:
... and now tests fail
Will take a look tomorrow

Reply:
Removed the increment again for now.

So in the case where we don't support seeking, `outout_position_` can be less than `position_`, since in that case only `position_` gets increased. Will in this case `output_position_` not be the position of the `external_output_processor_`?

Reply:
No, the position of the output processor is always incremented after a release_buffer, that is part of its definition.

" position: Represents the current position, in bytes, within the output
 * stream where the encoded data will be written next. This position moves
 * forward with each `release_buffer` call as data is written, and can also be
 * adjusted through the optional seek callback, if provided."

Reply:
Ok

In the 6 failing tests, the  "streaming_adapter.CheckFinalWatermarkPosition();" now fails. I suppose this means that at some place we are not calling `JxlEncoderOutputProcessorWrapper::SetFinalizedPosition` or somehting is wrong in FlushOutput, at some point we should call `output_processor.set_finalized_position` and don't.

wdyt, @veluca93 ?
`

Reply:
restructed it a bit, should work now.